keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,242 @@
+
+| Driver name       | value | Description  |
+| ----------------- | ----- | ------------ |
+| `ebs.csi.aws.com` | 0     | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
+| `ebs.csi.aws.com` | X>0   | plugin / CSI driver exists and can attach X volumes (where X > 0) |
+| `ebs.csi.aws.com` | X<0   | negative values are blocked by validation
+| key is missing in `CSINode.status.allocatable`   | -     |  there is no limit of volumes on the node* |
+
+*) This way we are not able to distinguish between a volume plugin / CSI driver not installed on a node or it has been installed and it has no limits.
+  * Predicates modified in this KEP assume that storage provided by **in-tree** volume plugin is available all nodes in the cluster. Other predicate(s) will evaluate `PV.spec.nodeAffinity` and filter out nodes that don't have access to the storage.
+  * For CSI drivers, availability of a CSI driver on a node can be checked in `CSINode.spec`.  
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+status:
+  allocatable:
+    # AWS node can attach max. 40 volumes, 1 is reserved for the system
+    ebs.csi.aws.com: 39
+```
+
+* Existing scheduler predicates `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are already deprecated.
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
+
+
+### New API
+
+CSINode gets `Status` struct with `Allocatable`, holding limit of volumes for each volume plugin and CSI driver that can be scheduled to the node.
+
+```go
+type CSINode struct {
+	...
+
+	// spec is the status of CSINode
+	Status CSINodeStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// VolumeLimits is map CSI driver name -> maximum count of volumes for the driver on the node.
+// For in-tree volume plugins, name of corresponding CSI driver is used.
+// Value can be either:
+// - Positive integer: that's the volume limit.
+// - Zero: such volume cannot be used on the node.
+// - key is missing in VolumeLimits: there is no volume limit, i.e. any number of volumes can be used on the node.
+type VolumeLimits map[string]int64
+
+// CSINodeStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeStatus struct {
+	// allocatable is a list of volume limits for each volume plugin and CSI driver on the node.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
```

Comment:
We should consider how future proof the API is. We may want to extend it not only to *count* attached volumes, but also sum their *size*. E.g. [typical GCE instances can handle only 64 TB of data](https://cloud.google.com/compute/docs/machine-types#predefined_machine_types). How would be such size limit expressed?
* `CSINode.Status.Allocatable["com.google.gce"] = 128; CSINode.Status.AllocatableSize["com.google.gce"] = "64TB"`
* Or `CSINode.Status.Allocatable["com.google.gce"] = VolumeLimits{Count: 128, Size: "64TB"}`

Reply:
I prefer the second so that all of a plugins limits are defined in one place.  That may change the "unlimited" vs 0 semantics a little

Reply:
We should also consider which one "scales" better while supporting backward compatibility. `map[string]struct` will probably work fine.

Reply:
Apparently maps of structs are not allowed, only lists: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

Reply:
Reworked as array of structs:

```
status:
  allocatable:
    - name: ebs.csi.aws.com
      volumes: 39
```